File: qylllll/短链接项目/前端/src/stores/links.js
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { linksAPI } from '@/api/links'

export const useLinksStore = defineStore('links', () => {
  const links = ref([])
  const isLoading = ref(false)
  const pagination = ref({
    current_page: 1,
    per_page: 20,
    total: 0,
    total_pages: 1
  })

  // 获取链接列表
  const fetchLinks = async (params = {}) => {
    isLoading.value = true
    try {
      const response = await linksAPI.getLinks(params)
      if (response.data?.success && response.data?.data) {
        links.value = response.data.data.links || []
        pagination.value = response.data.data.pagination || pagination.value
      } else {
        console.warn('API响应格式异常:', response.data)
        links.value = []
      }
    } catch (error) {
      console.error('获取链接列表失败:', error)
      // 确保在错误时也设置为空数组，避免 undefined
      links.value = []
    } finally {
      isLoading.value = false
    }
  }

  // 创建短链接
  const createLink = async (linkData) => {
    try {
      console.log('发送创建链接请求:', linkData)
      const response = await linksAPI.createLink(linkData)
      console.log('创建链接响应:', response.data)
      
      if (response.data?.success && response.data?.data) {
        // 确保 links.value 是数组
        if (Array.isArray(links.value)) {
          links.value.unshift(response.data.data)
        } else {
          links.value = [response.data.data]
        }
        return { success: true, data: response.data.data }
      }
    } catch (error) {
      console.error('创建链接失败:', error)
      
      // 详细的错误处理
      let errorMessage = '创建失败'
      
      if (error.response?.data) {
        const errorData = error.response.data
        console.log('错误响应数据:', errorData)
        
        // 处理422验证错误
        if (error.response.status === 422 && errorData.detail) {
          if (Array.isArray(errorData.detail)) {
            errorMessage = errorData.detail.map(err => {
              if (err.loc && err.msg) {
                return `${err.loc.join('.')}: ${err.msg}`
              }
              return err.msg || err
            }).join(', ')
          } else {
            errorMessage = errorData.detail
          }
        }
        // 其他错误
        else if (errorData.message) {
          errorMessage = errorData.message
        }
      }
      
      return { success: false, message: errorMessage }
    }
  }

  // 更新链接
  const updateLink = async (linkId, linkData) => {
    try {
      const response = await linksAPI.updateLink(linkId, linkData)
      if (response.data?.success && response.data?.data) {
        // 确保 links.value 是数组
        if (Array.isArray(links.value)) {
          const index = links.value.findIndex(link => link.id === linkId)
          if (index !== -1) {
            links.value[index] = response.data.data
          }
        }
        return { success: true, data: response.data.data }
      }
    } catch (error) {
      console.error('更新链接失败:', error)
      return { success: false, message: error.response?.data?.error?.message || '更新失败' }
    }
  }

  // 删除链接
  const deleteLink = async (linkId) => {
    try {
      const response = await linksAPI.deleteLink(linkId)
      if (response.data?.success) {
        // 确保 links.value 是数组
        if (Array.isArray(links.value)) {
          links.value = links.value.filter(link => link.id !== linkId)
        }
        return { success: true }
      }
    } catch (error) {
      console.error('删除链接失败:', error)
      return { success: false, message: error.response?.data?.error?.message || '删除失败' }
    }
  }

  // 切换链接状态
  const toggleLinkStatus = async (linkId, status) => {
    try {
      const response = await linksAPI.updateLinkStatus(linkId, status)
      if (response.data?.success) {
        // 确保 links.value 是数组
        if (Array.isArray(links.value)) {
          const index = links.value.findIndex(link => link.id === linkId)
          if (index !== -1) {
            links.value[index].status = status
          }
        }
        return { success: true }
      }
    } catch (error) {
      console.error('更新链接状态失败:', error)
      return { success: false, message: error.response?.data?.error?.message || '操作失败' }
    }
  }

  return {
    links,
    isLoading,
    pagination,
    fetchLinks,
    createLink,
    updateLink,
    deleteLink,
    toggleLinkStatus
  }
})

File: qylllll/短链接项目/后端/app/api/links.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import Optional

from ..database import get_db
from ..schemas import (
    LinkCreate, LinkUpdate, Link, LinkBatch, Response, PaginatedResponse
)
from ..crud import links as crud_links
from .auth import get_current_user
from ..models import User

router = APIRouter()


@router.post("/create", response_model=Response)
async def create_link(
    link: LinkCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """创建短链接"""
    try:
        db_link = crud_links.create_link(db=db, link=link, user_id=current_user.id)
        
        # 直接返回链接数据，包含short_url
        link_data = Link.from_orm(db_link).dict()
        link_data['short_url'] = f"http://localhost:8000/{db_link.short_code}"
        
        return Response(
            success=True,
            message="链接创建成功",
            data=link_data
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/batch", response_model=Response)
async def batch_create_links(
    links_data: LinkBatch,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """批量创建短链接"""
    try:
        db_links = crud_links.batch_create_links(db=db, urls=links_data.urls, user_id=current_user.id)
        
        links_result = []
        for db_link in db_links:
            links_result.append({
                "id": db_link.id,
                "original_url": db_link.original_url,
                "short_code": db_link.short_code,
                "short_url": f"http://localhost:8000/{db_link.short_code}"
            })
        
        return Response(
            success=True,
            message=f"成功创建 {len(db_links)} 个链接",
            data={"links": links_result}
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/", response_model=Response)
async def get_links(
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    search: Optional[str] = Query(None),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """获取用户的链接列表"""
    if search:
        links = crud_links.search_links(db, current_user.id, search, skip, limit)
        total = len(links)  # 简化实现
    else:
        links = crud_links.get_user_links(db, current_user.id, skip, limit)
        total = len(links)  # 实际应该查询总数
    
    links_data = [Link.from_orm(link).dict() for link in links]
    
    return Response(
        success=True,
        message="获取成功",
        data={
            "links": links_data,
            "pagination": {
                "current_page": skip // limit + 1,
                "per_page": limit,
                "total": total,
                "total_pages": (total + limit - 1) // limit
            }
        }
    )


@router.get("/{link_id}", response_model=Response)
async def get_link(
    link_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """获取单个链接详情"""
    link = crud_links.get_link(db, link_id)
    if not link or link.owner_id != current_user.id:
        raise HTTPException(status_code=404, detail="链接不存在")
    
    return Response(
        success=True,
        message="获取成功",
        data={"link": Link.from_orm(link).dict()}
    )


@router.put("/{link_id}", response_model=Response)
async def update_link(
    link_id: int,
    link_update: LinkUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """更新链接"""
    db_link = crud_links.update_link(db, link_id, link_update, current_user.id)
    if not db_link:
        raise HTTPException(status_code=404, detail="链接不存在")
    
    return Response(
        success=True,
        message="链接更新成功",
        data={"link": Link.from_orm(db_link).dict()}
    )


@router.delete("/{link_id}", response_model=Response)
async def delete_link(
    link_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """删除链接"""
    success = crud_links.delete_link(db, link_id, current_user.id)
    if not success:
        raise HTTPException(status_code=404, detail="链接不存在")
    
    return Response(
        success=True,
        message="链接删除成功"
    )


@router.get("/recent/list", response_model=Response)
async def get_recent_links(
    limit: int = Query(5, ge=1, le=20),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """获取最近创建的链接"""
    links = crud_links.get_recent_links(db, current_user.id, limit)
    links_data = [Link.from_orm(link).dict() for link in links]
    
    return Response(
        success=True,
        message="获取成功",
        data={"links": links_data}
    )


@router.get("/popular/list", response_model=Response)
async def get_popular_links(
    limit: int = Query(10, ge=1, le=50),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """获取热门链接"""
    links = crud_links.get_popular_links(db, current_user.id, limit)
    links_data = [Link.from_orm(link).dict() for link in links]
    
    return Response(
        success=True,
        message="获取成功",
        data={"links": links_data}
    )


@router.post("/{link_id}/toggle-status", response_model=Response)
async def toggle_link_status(
    link_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """切换链接状态（启用/禁用）"""
    link = crud_links.get_link(db, link_id)
    if not link or link.owner_id != current_user.id:
        raise HTTPException(status_code=404, detail="链接不存在")
    
    from ..schemas import LinkUpdate
    link_update = LinkUpdate(is_active=not link.is_active)
    db_link = crud_links.update_link(db, link_id, link_update, current_user.id)
    
    status_text = "启用" if db_link.is_active else "禁用"
    
    return Response(
        success=True,
        message=f"链接已{status_text}",
        data={"link": Link.from_orm(db_link).dict()}
    )


@router.get("/{link_id}/clicks", response_model=Response)
async def get_link_clicks(
    link_id: int,
    limit: int = Query(100, ge=1, le=1000),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """获取链接点击日志"""
    link = crud_links.get_link(db, link_id)
    if not link or link.owner_id != current_user.id:
        raise HTTPException(status_code=404, detail="链接不存在")
    
    clicks = crud_links.get_link_clicks(db, link_id, limit)
    
    from ..schemas import ClickLog
    clicks_data = [ClickLog.from_orm(click).dict() for click in clicks]
    
    return Response(
        success=True,
        message="获取成功",
        data={"clicks": clicks_data}
    )